### PR TITLE
feat(subgraph): Receipts improved

### PR DESCRIPTION
**Pull request by julien51**
# Description

We recently found that accessing the transaction's `value` was not accurate because the call to `pruchaseKeys` could be happening in a "larger" transaction with a different value.
In order to handle this, in the short term, we are looking at the GNPChange event's value.

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: subgraph/src/receipt.ts [hunk 1]
```diff
@@ -79,6 +80,22 @@ export function createReceipt(event: ethereum.Event): void {
...
     receipt.amountTransferred = event.transaction.value
+    // We cannot trust `event.transaction.value` because the purchase function could in fact
+    // be happening inside of a larger transaction whose value is not the amount transfered,
+    // In that case, we need to look up the GNPChanged event
+    if (logs) {
+      for (let i = 0; i < logs.length; i++) {
+        const txLog = logs[i]
+
+        if (
+          // txLog.address == UNLOCK_ADDRESS &&
```
Comment (julien51) [inline subgraph/src/receipt.ts:91]: I am not sure how to do that since we don't actually have Unlock's address here?

**Review by clemsos — APPROVED** [review]
LG but good to add a test (tests are https://github.com/unlock-protocol/unlock/blob/3fdb26c29bbdcaa47d3e3d5e67ce6016d9a29671/tests/test/subgraph.ts#L291

File: subgraph/src/receipt.ts [hunk 2]
```diff
@@ -79,6 +85,25 @@ export function createReceipt(event: ethereum.Event): void {
     log.debug('Creating receipt for base currency lock {}', [lockAddress])
     receipt.payer = event.transaction.from.toHexString()
     receipt.amountTransferred = event.transaction.value
+    // We cannot trust `event.transaction.value` because the purchase function could in fact
+    // be happening inside of a larger transaction whose value is not the amount transfered,
+    // In that case, we need to look up the GNPChanged event
+    if (logs) {
```
Comment (clemsos) [inline subgraph/src/receipt.ts:91]: It seems that part is only needed if the tx value is null

Comment (julien51) [reply subgraph/src/receipt.ts:91]: That's what I initially thought, but in fact the `transaction.value` could also be "wrong". Imagine if a safe tx buys 2 keys but starts with a 3rd tx in the buddle which funds the safe? Thent the value would be wrong

Comment (clemsos) [reply subgraph/src/receipt.ts:91]: Yes in the case of a safe or a multicall then the value can be calculated that way. Also for multiple purchases the value is the total of all keys purchased.

Comment (julien51) [reply subgraph/src/receipt.ts:91]: Actually, that is a very good point. For receipts we actually have a single receipt per transaction, even if it ends up minting multiple keys (similar to when you buy multiple things at once you only get a single receipt at the store).
So my approach of looking at the GNP is not great.

I guess I can loop over all the GNP changed events for now.

We _really_ need an event to be triggered for receipt, like `PurchaseReceipt()` that would include the required details: `payer`, `contract`, `totalPaid`, `hash`... etc.

File: subgraph/tests/mockTxReceipt.ts [hunk 3]
```diff
@@ -109,23 +109,66 @@ export function newTransactionReceipt(
...
     defaultAddressBytes
   )
 }
+
+// Create transaction receipt for mock transaction
+export function newCancelKeyTransactionReceipt(
+  tokenAddress: Address,
+  refund: BigInt
+): ethereum.TransactionReceipt {
+  return newTransactionReceipt([
```
Comment (julien51) [inline subgraph/tests/mockTxReceipt.ts:132]: Ohhh! That makes sense. I am not sure why I assumed there could be a single "log" per mocked transaction.

**Comment by clemsos** [comment]
after some digging, the integration tests are failing because the subgraph is enable to access the logs topics of the `GNPChanged` event. The log are correctly emitted but for some reasons, the event emitted by other contracts that the one tracked by the graph "datasource" is empty -no values in topics, just the `topics[0]` hash.

File: subgraph/src/receipt.ts [hunk 4]
```diff
@@ -76,9 +82,37 @@ export function createReceipt(event: ethereum.Event): void {
...
+      const lockContract = PublicLock.bind(Address.fromString(lockAddress))
+      const unlockAddress = lockContract.try_unlockProtocol()
+      let value = BigInt.zero()
+      for (let i = 0; i < logs.length; i++) {
+        const txLog = logs[i]
+        if (
+          txLog.address == unlockAddress.value &&
+          txLog.topics[0].toHexString() == GNP_CHANGED_TOPIC0
+        ) {
+          // log all topics
```
Comment (clemsos) [inline subgraph/src/receipt.ts:103]:
```suggestion
```

File: subgraph/tests/mockTxReceipt.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+// adds a GNPChanged event to the tx receipt
+export function newGNPChangedTransactionReceipt(
+  keyValue: BigInt,
+  totalValue: BigInt
+): ethereum.TransactionReceipt {
+  const eventSignature = Bytes.fromHexString(GNP_CHANGED_TOPIC0)
+  const grossNetworkProduct = BigInt.fromU32(0)
+
+  // as the tx is sent from another contract (NOT the lock)
+  // only the signature is passed as topic
```
Comment (clemsos) [inline subgraph/tests/mockTxReceipt.ts:159]:
```suggestion
  // only the event signature is passed as topics[0]
```

**Pull request merged**